{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MGCeek\n",
    "#### Take Specialized Metabolite gene pHMM hits (from (pl)anti-SMASH) and genomic gff to observe patterns in metabolic gene clustering   \n",
    "#### Author=Bob_Auber\n",
    "#### Date=07.08.2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this uses iPython magic to make plots appear inline\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.patches as patches\n",
    "import gzip\n",
    "import fileinput\n",
    "import glob\n",
    "import os\n",
    "from scipy import stats\n",
    "import re\n",
    "from collections import OrderedDict\n",
    "import statistics\n",
    "import seaborn as sns\n",
    "\n",
    "def runCMD(cmd):\n",
    "    val = subprocess.Popen(cmd, shell=True).wait()\n",
    "    if val == 0:\n",
    "        pass\n",
    "    else:\n",
    "        print ('command failed')\n",
    "        print (cmd)\n",
    "        sys.exit(1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Approach:\n",
    "## Go through genome\n",
    "### 1) Identitfy genes with SM HMM (SMG)\n",
    "### 2) Create a gene window X genes/basepairs up and downstream of each SMG\n",
    "### 3) Consolidate overlapping windows and observe instances of SMGs in windows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Rootdir='/depot/jwisecav/data/rauber/projects/prymnesium/genomes/2797_phase/v0.5/MGC/'\n",
    "gff='/depot/jwisecav/data/rauber/projects/prymnesium/genomes/2797_phase/v0.5/UTEX2797_genomic_v0.5.gff'\n",
    "window_size=50000 # this will capture genes that are X genes/basepairs upstream AND downstream of each SMG\n",
    "basepairs=True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define genes of interest to use as seeds for identifying gene windows of interest\n",
    "#### These genes are stored into gene_hmm_dict\n",
    "#### If using other gene sets, replace gene_hmm_dict with a dict of your choice in the format of 'geneID':[annotation1,annotation2,...]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Create dictionaries containing info of genes with SM hmms (SMGS)\n",
    "plant_out_files=glob.glob('/depot/jwisecav/data/rauber/projects/prymnesium/genomes/2797_phase/v0.5/annotation/hmms/plant/*hmm_out')\n",
    "fungal_out_files=glob.glob('/depot/jwisecav/data/rauber/projects/prymnesium/genomes/2797_phase/v0.5/annotation/hmms/fungal/*hmm_out')\n",
    "hmm_selection='all' #plant or fungi or KEGG or all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if hmm_selection == 'plant' or hmm_selection == 'all':\n",
    "    plant_gene_hmm_dict={}\n",
    "    for file in plant_out_files:\n",
    "        with open(file) as sources:\n",
    "            lines=sources.readlines()\n",
    "            for line in lines:\n",
    "                if line.startswith('UTEX') is True:\n",
    "                    #print(line)\n",
    "                    gene=line.split()[0].split('.')[0]\n",
    "                    #print(gene)\n",
    "                    evalue=float(line.split()[4])\n",
    "                    #print(evalue)\n",
    "                    HMM=line.split()[2]\n",
    "                    #print(HMM)\n",
    "                    \n",
    "                    \n",
    "                    if evalue < 1e-3:\n",
    "                        if gene in plant_gene_hmm_dict:\n",
    "                            plant_gene_hmm_dict[gene].append(HMM)\n",
    "                        else:\n",
    "                            plant_gene_hmm_dict[gene]=[HMM]\n",
    "                        \n",
    "if hmm_selection == 'fungi' or hmm_selection == 'all':                        \n",
    "    fungal_gene_hmm_dict={}\n",
    "    for file in fungal_out_files:\n",
    "        with open(file) as sources:\n",
    "            lines=sources.readlines()\n",
    "            for line in lines:\n",
    "                if line.startswith('UTEX') is True:\n",
    "                    #print(line)\n",
    "                    gene=line.split()[0].split('.')[0]\n",
    "                    #print(gene)\n",
    "                    evalue=float(line.split()[4])\n",
    "                    #print(evalue)\n",
    "                    HMM=line.split()[2]\n",
    "                    #print(HMM)\n",
    "                    \n",
    "                    \n",
    "                    if evalue < 1e-3:\n",
    "                        if gene in fungal_gene_hmm_dict:\n",
    "                            fungal_gene_hmm_dict[gene].append(HMM)\n",
    "                            \n",
    "                        else:\n",
    "                            fungal_gene_hmm_dict[gene]=[HMM]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1230 genes have a plant SM pHMM\n"
     ]
    }
   ],
   "source": [
    "print(len(plant_gene_hmm_dict), 'genes have a plant SM pHMM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1707 genes have a fungal SM pHMM\n"
     ]
    }
   ],
   "source": [
    "print(len(fungal_gene_hmm_dict), 'genes have a fungal SM pHMM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# construct non-redundant HMM dict dependent on plant or fungal input\n",
    "gene_hmm_dict={}\n",
    "if hmm_selection == 'plant' or hmm_selection == 'all':\n",
    "    #print('yay')\n",
    "    for gene in plant_gene_hmm_dict:\n",
    "        hmms=plant_gene_hmm_dict[gene]\n",
    "        for hmm in set(hmms):\n",
    "            if gene in gene_hmm_dict:\n",
    "                gene_hmm_dict[gene].append(hmm)\n",
    "            else:\n",
    "                gene_hmm_dict[gene]=[hmm]\n",
    "\n",
    "    \n",
    "if hmm_selection == 'fungi' or hmm_selection == 'all':\n",
    "    #print('yay')\n",
    "    for gene in fungal_gene_hmm_dict:\n",
    "        hmms=fungal_gene_hmm_dict[gene]\n",
    "        for hmm in set(hmms):\n",
    "            if gene in gene_hmm_dict:\n",
    "                gene_hmm_dict[gene].append(hmm)\n",
    "            else:\n",
    "                gene_hmm_dict[gene]=[hmm]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2400 genes have either a plant or fungal SM pHMM hit\n"
     ]
    }
   ],
   "source": [
    "print(len(gene_hmm_dict), 'genes have either a plant or fungal SM pHMM hit')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing of GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45535 genes read from gff file\n"
     ]
    }
   ],
   "source": [
    "#Generate gene:contig and gene:coordinate dictionaries \n",
    "gene_coord_dict={}\n",
    "gene_contig_dict={}\n",
    "\n",
    "count=0\n",
    "with open(gff, 'r') as sources:\n",
    "    lines=sources.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith('#') is False:\n",
    "            contig=line.split('\\t')[0]\n",
    "            feature=line.split('\\t')[2]\n",
    "            start=float(line.split('\\t')[3])\n",
    "            \n",
    "            if feature=='gene':\n",
    "                #I am pulling from the ID=gene in last gff column (first comma delimited info)\n",
    "                gene=line.split('\\t')[-1].split(';')[0].split('=')[1]\n",
    "                gene_coord_dict[gene]=start\n",
    "                gene_contig_dict[gene]=contig\n",
    "                count+=1\n",
    "print(count, 'genes read from gff file')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Identify genes +-X upstream and downstream of each SM gene and put in SMG_gene_window_dict\n",
    "\n",
    "SMG_gene_window_dict={}\n",
    "SMG_contig_window_dict={}\n",
    "\n",
    "#####EXTRACT COORDINATE INFO OF GENES######\n",
    "for SMG in gene_hmm_dict:\n",
    "    SMG_contig=gene_contig_dict[SMG]\n",
    "    SMG_start=gene_coord_dict[SMG]\n",
    "############################################\n",
    "    \n",
    "    ##########EXTRACT RELATIVE COORDINATES TO SMG##############\n",
    "    coordinates=[]\n",
    "    trimmed_coordinates=[]\n",
    "    relative_coord_dict={}\n",
    "    for gene in gene_contig_dict.keys():#for all genes\n",
    "        contig=gene_contig_dict[gene]\n",
    "        start=gene_coord_dict[gene]\n",
    "        \n",
    "        if contig == SMG_contig:\n",
    "            relative_coord=-(SMG_start-start)\n",
    "            \n",
    "            if relative_coord in relative_coord_dict.keys():\n",
    "                relative_coord_dict[relative_coord].append(gene)\n",
    "            else:\n",
    "                relative_coord_dict[relative_coord]=[gene]\n",
    "            coordinates.append(relative_coord)\n",
    "    coordinates.sort()       \n",
    "    #print(len(coordinates), 'genes were found on the contig containing', SMG)\n",
    "    ##########################################################\n",
    "    \n",
    "    \n",
    "    \n",
    "    ########FIND NUMBER OF GENES UPSTREAM AND DOWNSTREAM OF WINDOW############\n",
    "    ahead=0\n",
    "    behind=0\n",
    "    for coordinate in coordinates:\n",
    "        if basepairs is False:\n",
    "            if coordinate < 0:\n",
    "                behind+=1\n",
    "            if coordinate > 0:\n",
    "                ahead+=1\n",
    "\n",
    "        if basepairs is True:\n",
    "            if coordinate < -window_size:\n",
    "                behind+=1\n",
    "            if coordinate > window_size:\n",
    "                ahead+=1\n",
    "                \n",
    "    #print(behind, 'genes were behind', SMG)\n",
    "    #print(ahead, 'genes were ahead', SMG)\n",
    "    #############################################WARNING MESSAGES\n",
    "    #if basepairs is False:\n",
    "    #    if ahead < window_size:\n",
    "    #        print('WARNING! Size of gene window ahead', SMG, 'will be shortend to', ahead, 'genes instead of the specified window size:', window_size)\n",
    "    #    if behind < window_size:\n",
    "    #        print('WARNING! Size of gene window behind', SMG, 'will be shortend to', behind, 'genes instead of the specified window size:', window_size)\n",
    "    #if basepairs is True:\n",
    "    #    if max(coordinates) < window_size:\n",
    "    #        print('WARNING! Size of window ahead', SMG, 'will be shortend to', abs(max(coordinates)), 'basepairs instead of the specified window size:', window_size)\n",
    "    #    if min(coordinates) > -window_size:\n",
    "    #        print('WARNING! Size of window behind', SMG, 'will be shortend to', abs(min(coordinates)), 'basepairs instead of the specified window size:', window_size)     \n",
    "    ############################################\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    ###########WINDOW TRIMMING###############\n",
    "    if basepairs is False:\n",
    "        if ahead > window_size:\n",
    "            extra=ahead-window_size\n",
    "            #print(extra, 'extra genes ahead')\n",
    "            coordinates=coordinates[:-extra]\n",
    "    \n",
    "        if behind > window_size:\n",
    "            extra=behind-window_size\n",
    "            #print(extra, 'extra genes behind')\n",
    "            coordinates=coordinates[extra:]\n",
    "            \n",
    "    if basepairs is True:\n",
    "        if ahead > 0:\n",
    "            #print(extra, 'extra genes ahead')\n",
    "            coordinates=coordinates[:-ahead]\n",
    "        if behind > 0:\n",
    "            coordinates=coordinates[behind:]\n",
    "    #print(len(coordinates))\n",
    "    #######################################\n",
    "\n",
    "    \n",
    "    \n",
    "    ##########GENE WINDOW ASSOCIATIONS##########\n",
    "    for coordinate in coordinates:\n",
    "        genes=relative_coord_dict[coordinate]\n",
    "        for gene in genes:    \n",
    "            if SMG in SMG_gene_window_dict:\n",
    "                SMG_gene_window_dict[SMG].append(gene)\n",
    "            else:\n",
    "                SMG_gene_window_dict[SMG]=[gene]\n",
    "        ###########################################\n",
    "        \n",
    "     \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This dict contains each SMG and the genes up/downstream according to thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#SMG_gene_window_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Consolidate overlapping windows into one. (SMGs must be in same window to be combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "count=0\n",
    "raw_consolidated_SMG_gene_window_dict={}\n",
    "for SMG in SMG_gene_window_dict:\n",
    "    count+=1\n",
    "\n",
    "    raw_consolidated_SMG_gene_window_dict['window_'+str(count)]=set()\n",
    "    raw_consolidated_SMG_gene_window_dict['window_'+str(count)].add(SMG)\n",
    "    window_genes=SMG_gene_window_dict[SMG]\n",
    "    \n",
    "    #find all SMGS in window of SMG\n",
    "    for gene in window_genes:\n",
    "        if gene == SMG:\n",
    "            continue\n",
    "        else:\n",
    "            if gene in SMG_gene_window_dict.keys():\n",
    "                raw_consolidated_SMG_gene_window_dict['window_'+str(count)].add(gene)\n",
    "    \n",
    "   \n",
    "    \n",
    "    i=1\n",
    "    while i == 1: #now continually iterate through all SMGs in window to see if they have neighboring SMGs too\n",
    "        \n",
    "        \n",
    "        to_add_genes=set()\n",
    "        for window in raw_consolidated_SMG_gene_window_dict['window_'+str(count)]:\n",
    "            window_genes=SMG_gene_window_dict[window]\n",
    "            for gene in window_genes:\n",
    "                if gene in raw_consolidated_SMG_gene_window_dict['window_'+str(count)]: #if already identified in same window\n",
    "                    continue\n",
    "                if gene == SMG:\n",
    "                    continue\n",
    "                else:\n",
    "                    if gene in SMG_gene_window_dict.keys():\n",
    "                        to_add_genes.add(gene)\n",
    "                        \n",
    "        #now add new SMGs found next to neighboring SMGs\n",
    "        if len(to_add_genes) == 0:\n",
    "            i = 0\n",
    "            #print('done')\n",
    "        else:\n",
    "            #print(len(to_add_genes))\n",
    "            i=1 #iterate again\n",
    "            for gene in to_add_genes:\n",
    "                raw_consolidated_SMG_gene_window_dict['window_'+str(count)].add(gene)\n",
    "    \n",
    "                    \n",
    "    \n",
    "        \n",
    "        \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2400"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(raw_consolidated_SMG_gene_window_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Though windows are consolidated, they will be redudant for each overlapping SMG \n",
    "### Lets identitify them and only retain unique windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1281\n"
     ]
    }
   ],
   "source": [
    "overlap_sets=[]\n",
    "for window in raw_consolidated_SMG_gene_window_dict:\n",
    "    overlap_set=raw_consolidated_SMG_gene_window_dict[window]\n",
    "    if overlap_set in overlap_sets:\n",
    "        continue\n",
    "    else:\n",
    "        overlap_sets.append(overlap_set)\n",
    "print(len(overlap_sets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2400\n",
      "2400\n"
     ]
    }
   ],
   "source": [
    "overlap_genes=[]\n",
    "count=0\n",
    "for overlap_set in overlap_sets:\n",
    "    for gene in overlap_set:\n",
    "        count+=1\n",
    "        overlap_genes.append(gene)\n",
    "print(len(overlap_genes))\n",
    "print(len(set(overlap_genes)))#sanity check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "consolidated_SMG_gene_window_dict={}\n",
    "count=0\n",
    "for overlap_set in overlap_sets:\n",
    "    count+=1\n",
    "    consolidated_SMG_gene_window_dict['window_'+str(count)]=set()\n",
    "    for SMG in overlap_set:\n",
    "        window_genes=SMG_gene_window_dict[SMG]\n",
    "        for gene in window_genes:\n",
    "            consolidated_SMG_gene_window_dict['window_'+str(count)].add(gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1281 unique SMG windows found\n"
     ]
    }
   ],
   "source": [
    "print(len(consolidated_SMG_gene_window_dict), 'unique SMG windows found')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#consolidated_SMG_gene_window_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "691 ( 53.942232630757225 %) consolidated windows contain 1 SMGs\n",
      "314 ( 24.512099921935988 %) consolidated windows contain 2 SMGs\n",
      "146 ( 11.397345823575332 %) consolidated windows contain 3 SMGs\n",
      "74 ( 5.776736924277908 %) consolidated windows contain 4 SMGs\n",
      "27 ( 2.107728337236534 %) consolidated windows contain 5 SMGs\n",
      "15 ( 1.1709601873536302 %) consolidated windows contain 6 SMGs\n",
      "4 ( 0.312256049960968 %) consolidated windows contain 7 SMGs\n",
      "5 ( 0.39032006245121 %) consolidated windows contain 8 SMGs\n",
      "2 ( 0.156128024980484 %) consolidated windows contain 9 SMGs\n",
      "2 ( 0.156128024980484 %) consolidated windows contain 11 SMGs\n",
      "1 ( 0.078064012490242 %) consolidated windows contain 14 SMGs\n"
     ]
    }
   ],
   "source": [
    "#See how many SMGs are in each consolidated window\n",
    "SMG_consolidated_window_counts=[]\n",
    "SMG_consolidated_window_plot_dict={}\n",
    "\n",
    "for window in consolidated_SMG_gene_window_dict:\n",
    "\n",
    "    window_genes=consolidated_SMG_gene_window_dict[window]\n",
    "    count=0\n",
    "    for gene in window_genes:\n",
    "        if gene in gene_hmm_dict:\n",
    "            \n",
    "            count+=1\n",
    "    SMG_consolidated_window_counts.append(count)\n",
    "\n",
    "\n",
    "for threshold in range(0,max(SMG_consolidated_window_counts)+1):\n",
    "    count=0\n",
    "    for num in SMG_consolidated_window_counts:\n",
    "        if num == 0:\n",
    "            num += 1 \n",
    "        if num == threshold:\n",
    "            count+=1\n",
    "    if count != 0:\n",
    "        print(count, '(', (count/len(SMG_consolidated_window_counts))*100, '%) consolidated windows contain' , threshold ,'SMGs')\n",
    "        SMG_consolidated_window_plot_dict[str(threshold)]=count\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example how to write output to files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#with open(Rootdir + '10_gene_all_hmm_and_KEGG_SMG_consolidated_windows.txt', 'w') as f:\n",
    "#    for window in consolidated_SMG_gene_window_dict:\n",
    "#        window_genes=consolidated_SMG_gene_window_dict[window]\n",
    "#        f.write(window+':'+','.join(window_genes)+'\\n')\n",
    "        \n",
    "#with open(Rootdir + '50kb_all_hmm_SMG_consolidated_windows.txt', 'w') as f:\n",
    "#    for window in consolidated_SMG_gene_window_dict:\n",
    "#        window_genes=consolidated_SMG_gene_window_dict[window]\n",
    "#        f.write(window+':'+','.join(window_genes)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (env.genomics kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
